src/lib/gadgets/foreign-field.ts

```diff
@@ -92,12 +107,230 @@ function singleAdd(x: Field3, y: Field3, sign: Sign, f: bigint) {
   return { result: [r0, r1, r2] satisfies Field3, overflow };
 }
 
+function multiply(a: Field3, b: Field3, f: bigint): Field3 {
+  assert(f < 1n << 259n, 'Foreign modulus fits in 259 bits');
+
+  // constant case
+  if (a.every((x) => x.isConstant()) && b.every((x) => x.isConstant())) {
+    let ab = Field3.toBigint(a) * Field3.toBigint(b);
+    return Field3.from(mod(ab, f));
+  }
+
+  // provable case
+  let { r01, r2, q, q2Bound } = multiplyNoRangeCheck(a, b, f);
+
+  // limb range checks on quotient and remainder
+  multiRangeCheck(q);
+  let r = compactMultiRangeCheck(r01, r2);
+
+  // range check on q and r bounds
+  // TODO: this uses one RC too many.. need global RC stack, or get rid of bounds checks
```

Comment:
fwiw I think we can get rid of the bounds checks entirely if we just slightly tweak the rangecheck1 gate. I'll do this after the HF.

therefore, I don't think it's worth introducing the global variable storing pending limbs. this would complicate the architecture, and the impact is pretty small: in the ec addition gadget which takes 70 rows, this would currently only save us 2/3s of a MRC gate, so only 2.66 rows. That's just a 3.8% improvement. I still have a couple of ideas for improvements that are much more impactful and don't complicate the overall architecture.

Reply:
Thanks for asking about the exact layout and giving an example @querolita. It made me realize a mistake I made -- I accidentally range-checked the carry c0, even though it is just 2 bits and is checked in the ffmul gate 😬